ssl/statem/extensions.c

```diff
@@ -939,9 +939,15 @@ static int final_server_name(SSL *s, unsigned int context, int sent)
      */
     if (s->server) {
         if (!sent) {
-            /* Nothing from the client this handshake; cleanup stale value */
-            OPENSSL_free(s->ext.hostname);
-            s->ext.hostname = NULL;
+            /*
+             * Nothing used from the client this handshake; from the
+             * perspective of TLS any value here is stale/unused.  However,
+             * we have a de facto stable ABI guarantee to preserve this value
+             * for subsequent retrieval with SSL_get_servername() even if it
+             * was not used to negotiate an SNI value for this session.
+             * TODO(OpenSSL1.2) cleanup stale value.
+             */
```

Comment:
Why would there have been a "stale value" in the SSL handle in the first place? If the client did not send the extension, what does this value represent???

Reply:
I think there are renegotiation flows where there could be a non-NULL value here even if the client did not sent SNI in the current ClientHello.  (#6378 should have more discussion on this point, IIRC)